src/java.base/share/classes/sun/security/provider/SHA3.java

```diff
@@ -91,10 +95,12 @@ void implCompress(byte[] b, int ofs) {
 
     @IntrinsicCandidate
     private void implCompress0(byte[] b, int ofs) {
-       for (int i = 0; i < buffer.length; i++) {
-           state[i] ^= b[ofs++];
-       }
-       keccak();
+        b2lLittle(b, ofs, longBuf, 0, blockSize);
+        for (int i = 0; i < blockSize / 8; i++) {
+            state[i] ^= longBuf[i];
```

Comment:
Clever. So the intrinsic (C2 code) still generates code corresponding original loop with `byte b[]` array. This will be confusing. It will also slowdown execution in Interpreter so - additional array copy.

New code also assumes that `buffer.length == blockSize` and `(buffer.length % 8) == 0`. I hope there is some assertions/checks in java code to verify that.

Some one from core-libs have to review this.

Reply:
Well, the intrinsic function treats the input and state as long arrays anyways, and so it only works on little endian architectures, where the conversion is a no-op. There is no additional array copy, this b2lLittle() call used to be in the keccak() method (along with the conversion back to byte array), the point of this whole change is that only one of these conversions should be done with every keccak() call (an additional benefit is that the xor and the corresponding loads+store is done on longs, not on bytes).

Reply:
Oh, and about the length: buffer is allocated in the constructor of the parent class (DigestBase) like this:
  buffer = new byte[blockSize];
Here blockSize is one of { 72, 104, 136, 144, 168 }, so divisible by 8.
buffer.length was used before probably because blockSize was declared private in DigestBase. I made it protected, because in my opinion it is easier to read the code this way.

Reply:
Thank you for explanation.

An other question. Is any other use of `longBuf` array after `implCompress0()` call ? Because Intrinsic code will not update it.

Reply:
There is no other use.. That array is there for the sole purpose of  holding the input as a long array before it gets xor-ed into the state.

Reply:
okay

Reply:
Could you try using MethodHandles instead of b2lLittle? Similar to what's used in ChaCha20:
https://github.com/openjdk/jdk/blob/b77bd5fd6a6f7ddbed90300fba790da4fb683275/src/java.base/share/classes/com/sun/crypto/provider/ChaCha20Cipher.java#L132-L134

with MethodHandles you could drop longBuf, removing unnecessary allocation on the only platform where the intrinsic is implemented (MacOSX-aarch64)

Reply:
b2lLittle() and l2bLittle() are using those MethodHandle stuff, but yes, it adds some more performance if I use them directly in this class, so I changed it.